File: frontend/types/vacaciones.ts
export type EstadoVacacion = "planificada" | "tomada" | "cancelada"
export type TipoVacacion = "vacaciones" | "semana_free" | "dia_free" | "permiso_especial"

export interface SolicitudVacaciones {
  id: string
  empresa_id: string
  empresa_nombre: string | null
  empleado_id: string
  empleado_nombre: string | null
  area_id: string | null
  area_nombre: string | null
  fecha_desde: string  // ISO date "YYYY-MM-DD"
  fecha_hasta: string  // ISO date "YYYY-MM-DD"
  dias: number
  tipo: TipoVacacion
  comentario: string | null
  cancelada: boolean
  estado: EstadoVacacion  // derivado por el backend
  /** Año al que CORRESPONDE la licencia; puede diferir del año de fecha_desde. */
  periodo: number | null
  dias_liquidados: number
  created_at: string
}

export interface SolicitudVacacionesCreate {
  empleado_id: string
  fecha_desde: string
  fecha_hasta: string
  tipo?: TipoVacacion
  comentario?: string
  periodo?: number
  dias_liquidados?: number
}

export interface SolicitudVacacionesUpdate {
  comentario?: string
  tipo?: TipoVacacion
  periodo?: number
  dias_liquidados?: number
}

/**
 * Días de un período que NO se tomaron. Viven en otra tabla porque no tienen fechas
 * (nadie faltó ningún día) — el porqué está en backend/migrations/083.
 *
 * `dias_liquidados` es un entero y no un bool: admite liquidación parcial. La UI lo muestra
 * como un tilde (tildado → dias_liquidados = dias).
 */
export interface VacacionPendiente {
  id: string
  empresa_id: string
  empresa_nombre: string | null
  empleado_id: string
  empleado_nombre: string | null
  area_id: string | null
  area_nombre: string | null
  periodo: number
  dias: number
  dias_liquidados: number
  comentario: string | null
  created_at: string
}

export interface VacacionPendienteCreate {
  empleado_id: string
  periodo: number
  dias: number
  dias_liquidados?: number
  comentario?: string
}

export interface VacacionPendienteListResponse {
  items: VacacionPendiente[]
  total: number
}

export interface SolicitudVacacionesListResponse {
  items: SolicitudVacaciones[]
  total: number
}

export interface SaldoPeriodo {
  periodo: number
  cupo: number
  gozados: number
  pedidos: number
  disponibles: number
  /** ISO "YYYY-MM-DD". Se formatea partiendo el string, NUNCA con `new Date()`: esa fecha se
   *  parsea en UTC y en Argentina (UTC−3) se muestra un día antes — el 31/12 se ve como 30/12,
   *  que en un vencimiento es el día que importa. */
  vence: string
  vencido: boolean
}

export interface SaldoVacaciones {
  empleado_id: string
  /** Cupo calculado de los períodos NO vencidos — no es la columna `dias_vacaciones_asignados`,
   *  que desde la migración 090 es un override opcional. */
  asignados: number
  gozados: number
  pedidos: number
  disponibles: number
  vencidos: number
  /** Opcional en el TIPO a propósito: un backend viejo (o un 200 cacheado) no lo manda, y el
   *  componente tiene que seguir mostrando los cuatro totales en vez de romper. */
  por_periodo?: SaldoPeriodo[]
}

File: backend/schemas/vacaciones.py
"""
Schemas Pydantic para el módulo de vacaciones.
SolicitudVacacionesCreate → Update → Response → ListResponse

El campo `estado` en Response es DERIVADO (calculado en el service, no en la DB):
  - "cancelada"   si cancelada=True
  - "planificada" si hoy < fecha_desde
  - "tomada"      en cualquier otro caso (en curso o pasada)

Tipos de solicitud: vacaciones | semana_free | dia_free | permiso_especial
Solo 'vacaciones' descuenta del saldo anual del empleado.
"""
from datetime import date, datetime
from typing import List, Optional
from uuid import UUID

from pydantic import BaseModel, field_validator

TIPOS_VALIDOS = {"vacaciones", "semana_free", "dia_free", "permiso_especial"}


class SolicitudVacacionesCreate(BaseModel):
    empleado_id: UUID
    fecha_desde: date
    fecha_hasta: date
    comentario: Optional[str] = None
    tipo: str = "vacaciones"
    # Año al que CORRESPONDE la licencia, que puede diferir del año de fecha_desde: una
    # licencia tomada en 2026 puede ser del período 2025. Opcional — las filas previas a la
    # migración 083 no lo tienen y no se puede derivar (ver el encabezado de esa migración).
    periodo: Optional[int] = None
    dias_liquidados: int = 0

    @field_validator("periodo")
    @classmethod
    def validate_periodo(cls, v: Optional[int]) -> Optional[int]:
        if v is not None and not (2000 <= v <= 2100):
            raise ValueError(f"periodo inválido '{v}'. Rango: 2000–2100")
        return v

    @field_validator("tipo")
    @classmethod
    def validate_tipo(cls, v: str) -> str:
        if v not in TIPOS_VALIDOS:
            raise ValueError(f"tipo inválido '{v}'. Valores válidos: {sorted(TIPOS_VALIDOS)}")
        return v


class SolicitudVacacionesUpdate(BaseModel):
    comentario: Optional[str] = None
    tipo: Optional[str] = None
    periodo: Optional[int] = None
    dias_liquidados: Optional[int] = None

    @field_validator("tipo")
    @classmethod
    def validate_tipo(cls, v: Optional[str]) -> Optional[str]:
        if v is not None and v not in TIPOS_VALIDOS:
            raise ValueError(f"tipo inválido '{v}'. Valores válidos: {sorted(TIPOS_VALIDOS)}")
        return v


class SolicitudVacacionesResponse(BaseModel):
    id: str
    empresa_id: str
    empresa_nombre: Optional[str] = None
    empleado_id: str
    empleado_nombre: Optional[str] = None
    area_id: Optional[str] = None
    area_nombre: Optional[str] = None
    fecha_desde: date
    fecha_hasta: date
    dias: int
    tipo: str
    comentario: Optional[str] = None
    cancelada: bool
    estado: str  # "planificada" | "tomada" | "cancelada"
    periodo: Optional[int] = None      # año devengado; None en las filas previas a la mig. 083
    dias_liquidados: int = 0
    created_at: datetime


class SolicitudVacacionesListResponse(BaseModel):
    items: List[SolicitudVacacionesResponse]
    total: int


class SaldoVacacionesResponse(BaseModel):
    empleado_id: str
    asignados: int
    gozados: int   # tomadas (estado="tomada", tipo="vacaciones")
    pedidos: int   # planificadas (estado="planificada", tipo="vacaciones")
    disponibles: int  # asignados − gozados − pedidos
